### PR TITLE
Update the convert (match) function to CIEDE2000 using color-diff

### LINKED ISSUE #6: Improve color conversion accuracy
Currently the algorithm we use to convert colors is far from ideal. We should work on making this more accurate (which is highly subjective and we'll need to discuss this further)

### PR DESCRIPTION
**Pull request by evelynhathaway**
- Adds `color-diff` as a dependency for CIEDE2000
- Rewrote `convert()`
- Added utility function and type for uppercasing tinycolor's objects which should be removed after contributing to color-diff

Fixes #6

### CODE REVIEW

File: src/shared/convert.js [hunk 1]
```diff
@@ -1,35 +1,49 @@
...
+
+export function convert(fromHex: string): ColorMatchType {
+  // Convert input color to upcase RGB object
+  const fromRgb = hexToUpcaseRgb(fromHex);
+
+  if (!colorMapRgb) {
+    // Create `Map` with pairs of `upcaseRgb: paletteColorObject`
+    colorMapRgb = new Map(
+      colorArray.map((colorObject) => [hexToUpcaseRgb(colorObject.value), colorObject])
+    );
```
Comment (crutchcorn) [inline src/shared/convert.js]: I'm left unclear from initial read why a `Map` is used here as opposed to a `reduce` + Object? Does the ordering of the `map` somehow increase the speed of the `diff.closest`?

If there's no performance gain and you're okay with it, I'd rather have a reduce as I find the `Map` implementation a little unclear personally 😅 (not due to anything of yours, just my silly mental blocks)

If we were to, would could redo this implementation as:
```javascript
const {obj, arr} = colorArray.reduce((prev, colorObject) => {
   const hexUpcaseRgbObjVal = hexToUpcaseRgb(colorObject.value);
   prev.obj[hexUpcaseRgbObjVal] = colorObject;
   prev.arr.push(hexUpcaseRgbObjVal);
   return prev;
}, {obj: {}, arr: []});

palette = arr;
colorMapRgb = obj;
```

And even knock out doing the `Array.from` later.

File: src/shared/convert.js [hunk 2]
```diff
@@ -1,35 +1,49 @@
...
-}
+// Hoist upcase RGB objects so they're only created once (on first use)
+let colorMapRgb: Map<UpcaseRgbType, ColorMatchType>;
+let palette: Array<UpcaseRgbType>;
+
+export function convert(fromHex: string): ColorMatchType {
+  // Convert input color to upcase RGB object
+  const fromRgb = hexToUpcaseRgb(fromHex);
+
+  if (!colorMapRgb) {
```
[the remark this answers is not in the document]
Comment (evelynhathaway) [reply src/shared/convert.js]: Agreed. I originally avoided `&&` for the same reason I like being more clear about making sure both exist.

File: src/shared/convert.js [hunk 3]
```diff
@@ -1,35 +1,49 @@
...
+// `color-diff` only allows uppercase {R,G,B}
+// TODO: remove upcasing after submitting a PR for downcase
+type UpcaseRgbType = {
+  R: number,
+  G: number,
+  B: number
+};
+const hexToUpcaseRgb = function (hex: string): UpcaseRgbType {
+  const rgbColor = tinycolor(hex).toRgb();
+  return {R: rgbColor.r, G: rgbColor.g, B: rgbColor.b};
```
[the remark this answers is not in the document]
Comment (evelynhathaway) [reply src/shared/convert.js]: Sure, I like your style. I really wished `return {r ad R, g as G, b as B} = tinycolor(hex).toRgb();` was valid somehow haha.

[the remark this answers is not in the document]
Comment (crutchcorn) [reply src/shared/convert.js]: You can do this:
```javascript
const {r: R, g: G, b: B} = tinycolor(hex).toRgb();
return {R, G, B}
```

If you like that better (I do personally but figured you might not)

File: src/shared/convert.js (hunk 1 above)
Comment (crutchcorn) [reply src/shared/convert.js]: @evelynhathaway just explained to me that 
> `hexUpcaseRgbObjVal` is an object so it cannot be a key which is why I made a Map 

Which makes perfect sense.

<details>
This is why we don't code review at night, folks
</details>

Comment (crutchcorn) [inline src/shared/convert.js]: Let's add a comment that explains that `hexToUpcase` is not a string

File: src/shared/convert.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (evelynhathaway) [reply src/shared/convert.js]: Wait, what?! I love this but need to figure out why it works like that so I can read it more fluently.

[the remark this answers is not in the document]
Comment (crutchcorn) [reply src/shared/convert.js]: Admittedly the `{l: L}` (lowercase: UPPERCASE) does not help the readability here.

[It's part of the object destructuing syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Assigning_to_new_variable_names).

Then we just return the variables as their own objects using the [object shorthand property assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#New_notations_in_ECMAScript_2015)

[the remark this answers is not in the document]
Comment (evelynhathaway) [reply src/shared/convert.js]: I specifically didn't know that you could destructure to other variables. I love it and am gonna use it more often now.

File: src/shared/convert.js (hunk 1 above)
Comment (evelynhathaway) [reply src/shared/convert.js]: I modified a couple of comments to specifically mention their type while mentioning that they're objects. 😄

**Comment by evelynhathaway** [comment]
Pushed the suggestions 😃

**Pull request merged**